test/parallel/test-vm-dynamic-import.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+// Flags: --experimental-vm-modules --experimental-modules
+
+const common = require('../common');
+const assert = require('assert');
+common.crashOnUnhandledRejection();
+
+const vm = require('vm');
+
+async function testModule() {
+  const module = new vm.Module('import("X")');
+  await module.link(common.mustCall((specifier) => {
+    assert.strictEqual(specifier, 'X');
+    return new vm.Module('export const x = 5');
+  }));
+  module.instantiate();
+  const { result } = await module.evaluate();
+  assert.strictEqual((await result).x, 5);
+}
+
+async function testScript() {
+  {
+    let moduleX;
+    const script = new vm.Script('import("vm://X")', {
+      resolveDynamicImport(specifier, scriptOrModule) {
+        if (specifier === 'vm://X') {
+          assert.strictEqual(scriptOrModule, script);
+          return moduleX = new vm.Module('export default import("vm://Y")', { url: 'vm://X' });
+        }
+        if (specifier === 'vm://Y') {
+          assert.strictEqual(scriptOrModule, moduleX);
+          return new vm.Module('export default 5', { url: 'vm://Y' });
+        }
+      },
+    });
+    const result = script.runInThisContext();
+    assert.strictEqual((await (await result).default).default, 5);
+  }
+
+  {
+    const ns = {};
+    const script = new vm.Script('import("X")', {
+      resolveDynamicImport: common.mustCall(() => ns),
+    });
+    const result = script.runInThisContext();
+    assert.strictEqual(await result, ns);
```

Comment:
Thanks, this is what I wanted to catch - this must definitely be a failure, otherwise we aren't complying to spec.

The return value of resolveDynamicImport must either be a namespace object or a vm.Module object right?

So is the issue specifically is that v8 doesn't provide a Value.IsModuleNamespace implementation?

If so, perhaps this is something the API needs to expose.

For now a `obj[Symbol.toStringTag] === 'Module'` can work as a check at least.

Reply:
there are a lot of reasons why the object might not be an actual module namespace, our builtin named exports as one example. and fwiw "isModuleNamespace" would have to just check toStringTag because its just a sealed object with a toStringTag and a null prototype